src/sidebar/store/modules/test/groups-test.js

```diff
@@ -108,4 +139,140 @@ describe('sidebar.store.modules.groups', () => {
       assert.equal(store.focusedGroupId(), privateGroup.id);
     });
   });
+
+  describe('getFeaturedGroups', () => {
+    [
+      {
+        description:
+          'If logged in and there is a restricted group of which the user is not a member show it in `Featured Groups`.',
+        isLoggedIn: true,
+        filteredGroups: [restrictedGroup],
+        expectedGroups: [restrictedGroup],
+      },
+      {
+        description:
+          'If logged in and there is an open group show it in `Featured Groups`.',
+        isLoggedIn: true,
+        filteredGroups: [openGroup],
+        expectedGroups: [openGroup],
+      },
+      {
+        description:
+          'If logged in and there are only groups with membership do not show `Featured Groups`.',
+        isLoggedIn: true,
+        filteredGroups: [publicGroup],
+        expectedGroups: [],
+      },
+      {
+        description:
+          'If logged out and there is a restricted group scoped to the page show it in `Featured Groups`.',
+        isLoggedIn: false,
+        filteredGroups: [restrictedGroup],
+        expectedGroups: [restrictedGroup],
+      },
+      {
+        description:
+          'If logged out and there is an open group show it in `Featured Groups`.',
+        isLoggedIn: false,
+        filteredGroups: [openGroup],
+        expectedGroups: [openGroup],
+      },
+    ].forEach(({ description, isLoggedIn, filteredGroups, expectedGroups }) => {
+      it(description, function() {
+        store.getState().session = { userid: isLoggedIn ? '1234' : null };
+        store.loadGroups(filteredGroups);
+
+        const featuredGroups = store.getFeaturedGroups();
+        const currentlyViewing = store.getCurrentlyViewingGroups();
+
+        assert.isTrue(featuredGroups.length === expectedGroups.length);
+        let i;
+        for (i = 0; i < featuredGroups.length; ++i) {
+          assert.isTrue(featuredGroups[i].id === expectedGroups[i].id);
+        }
+        assert.isTrue(currentlyViewing.length === 0);
```

Comment:
The currently-viewing test statement here feels out of place in a unit-ish test...

Reply:
I wanted to be sure I covered all exclusive cases (where a group didn't get dropped and only showed up in one group list at a time). The CurrentlyViewing tests do a good job at covering the cases where groups show up in CurrentlyViewing and not in the other lists but if I remove this there won't be any tests to make sure CurrentlyViewing is empty when the other group lists contain groups. 
I just ran some mutation testing and there's no test to catch the CurrentlyViewing returning groups that are also in the Featured list for example.

Reply:
Interesting discussion. I generally subscribe to the principle that, where possible, it is useful to choose a representation for data such that only valid states can be represented. For example, if you have an ordered list of things and there should be no duplicates, use an ordered set rather than a list if possible. In this case you could do that by having a function which mapped group IDs to the list the group appeared in. However, that might not be convenient to consume in the UI. I guess that's why you went with three helper functions?

An alternative, which will also work fine here, might be to have a helper function that returns one of the lists and also checks the invariant that each group only appears in one list.

```js
// getList(...) retrieves the list and also checks that no groups appear in multiple lists.
assert.deepEqual(getList('featured'), expectedList)
```

Checking for dupes in an array can be done with:

```js
const hasDuplicates = array => new Set(array).size < array.length
```